chain/datasource/datasource.go

```diff
@@ -376,104 +394,104 @@ func GetGenesisActors(ctx context.Context, store adt.Store, genesis *types.TipSe
-	changes, err := diff.Hamt(ctx, oldR, newR, store, store, hamt.UseTreeBitWidth(5), hamt.UseHashFunction(func(input []byte) []byte {
+func fastDiff(ctx context.Context, store adt.Store, oldTree, newTree *StateTreeMeta) (tasks.ActorStateChangeDiff, error) {
+	oldMap, err := oldTree.LoadMap(store)
+	if err != nil {
+		return nil, err
+	}
+	newMap, err := newTree.LoadMap(store)
+	if err != nil {
+		return nil, err
+	}
+	changes, err := diff.Hamt(ctx, oldMap, newMap, store, store, hamt.UseTreeBitWidth(builtin.DefaultHamtBitwidth), hamt.UseHashFunction(func(input []byte) []byte {
 		res := sha256.Sum256(input)
 		return res[:]
 	}))
 	if err != nil {
 		return nil, err
 	}
+
 	buf := bytes.NewReader(nil)
 	out := map[address.Address]tasks.ActorStateChange{}
+
 	for _, change := range changes {
 		addr, err := address.NewFromBytes([]byte(change.Key))
 		if err != nil {
 			return nil, fmt.Errorf("address in state tree was not valid: %w", err)
 		}
+
 		var ch tasks.ActorStateChange
+
 		switch change.Type {
 		case hamt.Add:
 			ch.ChangeType = tasks.ChangeTypeAdd
-			buf.Reset(change.After.Raw)
-			err = ch.Actor.UnmarshalCBOR(buf)
-			buf.Reset(nil)
-			if err != nil {
-				return nil, err
+			if newTree.Tree.Version() <= types.StateTreeVersion4 {
+				var act types.ActorV4
+				buf.Reset(change.After.Raw)
+				err := act.UnmarshalCBOR(buf)
+				buf.Reset(nil)
+				if err != nil {
+					return nil, err
+				}
+				ch.Actor = *types.AsActorV5(&act)
+			} else {
+				var act types.Actor
+				buf.Reset(change.After.Raw)
+				err := act.UnmarshalCBOR(buf)
+				buf.Reset(nil)
+				if err != nil {
+					return nil, err
+				}
+				ch.Actor = act
 			}
+
 		case hamt.Remove:
-			ch.ChangeType = tasks.ChangeTypeRemove
-			buf.Reset(change.Before.Raw)
-			err = ch.Actor.UnmarshalCBOR(buf)
-			buf.Reset(nil)
-			if err != nil {
-				return nil, err
+			if newTree.Tree.Version() <= types.StateTreeVersion4 {
+				var act types.ActorV4
+				buf.Reset(change.Before.Raw)
+				err := act.UnmarshalCBOR(buf)
+				buf.Reset(nil)
+				if err != nil {
+					return nil, err
+				}
+				ch.Actor = *types.AsActorV5(&act)
+			} else {
+				var act types.Actor
+				buf.Reset(change.Before.Raw)
+				err := act.UnmarshalCBOR(buf)
+				buf.Reset(nil)
+				if err != nil {
+					return nil, err
+				}
+				ch.Actor = act
 			}
+
 		case hamt.Modify:
-			ch.ChangeType = tasks.ChangeTypeModify
-			buf.Reset(change.After.Raw)
-			err = ch.Actor.UnmarshalCBOR(buf)
-			buf.Reset(nil)
-			if err != nil {
-				return nil, err
+			if newTree.Tree.Version() <= types.StateTreeVersion4 {
+				var act types.ActorV4
+				buf.Reset(change.After.Raw)
+				err := act.UnmarshalCBOR(buf)
+				buf.Reset(nil)
+				if err != nil {
+					return nil, err
+				}
+				ch.Actor = *types.AsActorV5(&act)
+			} else {
+				var act types.Actor
+				buf.Reset(change.After.Raw)
+				err := act.UnmarshalCBOR(buf)
+				buf.Reset(nil)
+				if err != nil {
+					return nil, err
+				}
+				ch.Actor = act
```

Comment:
This is the actual fix; we address it by mimicking the behaviour of lotus: https://github.com/filecoin-project/lotus/blob/v1.20.0-rc2/chain/state/statetree.go#L631